pkg/kubectl/cmd/completion.go

```diff
@@ -0,0 +1,108 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"context"
+	"strings"
+	"time"
+
+	"github.com/spf13/cobra"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+	"k8s.io/kubectl/pkg/cmd/util"
+)
+
+type filterFunc func(config clientcmdapi.Config, toComplete string) []string
+
+func getFromConfig(f util.Factory, filter filterFunc) func(*cobra.Command, []string, string) ([]string, cobra.ShellCompDirective) {
+	return func(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+		config, err := f.ToRawKubeConfigLoader().RawConfig()
+		if err != nil {
+			return nil, cobra.ShellCompDirectiveNoFileComp
+		}
+		return filter(config, toComplete), cobra.ShellCompDirectiveNoFileComp
+	}
+}
+
+func registerCompletionFuncForGlobalFlags(cmds *cobra.Command, f util.Factory) error {
+	err := cmds.RegisterFlagCompletionFunc("namespace", func(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+		clientSet, err := f.KubernetesClientSet()
+		if err != nil {
+			return nil, cobra.ShellCompDirectiveDefault
+		}
+		ctx, cancel := context.WithTimeout(context.TODO(), time.Second*3)
+		defer cancel()
+		namespaces, err := clientSet.CoreV1().Namespaces().List(ctx, metav1.ListOptions{})
+		if err != nil {
+			return nil, cobra.ShellCompDirectiveDefault
+		}
+		var ret []string
+		for _, ns := range namespaces.Items {
+			if strings.HasPrefix(ns.Name, toComplete) {
+				ret = append(ret, ns.Name)
+			}
+		}
+		return ret, cobra.ShellCompDirectiveNoFileComp
+	})
+	if err != nil {
+		return err
+	}
+
+	getContextFromConfig := getFromConfig(f, func(config clientcmdapi.Config, toComplete string) []string {
+		var ret []string
+		for name := range config.Contexts {
+			if strings.HasPrefix(name, toComplete) {
+				ret = append(ret, name)
+			}
+		}
+		return ret
+	})
+	getClusterFromConfig := getFromConfig(f, func(config clientcmdapi.Config, toComplete string) []string {
+		var ret []string
+		for name := range config.Clusters {
+			if strings.HasPrefix(name, toComplete) {
+				ret = append(ret, name)
+			}
+		}
+		return ret
+	})
+	getUserFromConfig := getFromConfig(f, func(config clientcmdapi.Config, toComplete string) []string {
+		var ret []string
+		for name := range config.AuthInfos {
+			if strings.HasPrefix(name, toComplete) {
+				ret = append(ret, name)
+			}
+		}
+		return ret
+	})
+
+	err = cmds.RegisterFlagCompletionFunc("context", getContextFromConfig)
+	if err != nil {
```

Comment:
I'm not familiar with `CheckErr` but would it be more instructive to call it in this function instead of returning the error?

Reply:
Yes, I think you are right.